src/counter.rs

```diff
@@ -0,0 +1,42 @@
+use std::sync::atomic::{AtomicIsize, Ordering};
+
+pub(crate) struct ConcurrentCounter {
+    base: AtomicIsize,
+    cells: Vec<AtomicIsize>,
+}
+
+impl ConcurrentCounter {
+    pub(crate) fn new() -> Self {
+        Self {
+            base: AtomicIsize::new(0),
+            cells: (0..crate::map::num_cpus())
+                .into_iter()
+                .map(|_| AtomicIsize::new(0))
+                .collect(),
+        }
+    }
+
+    pub(crate) fn add(&self, value: isize) {
+        let base = self.base.load(Ordering::SeqCst);
+
+        while self
+            .base
+            .compare_exchange(base, base + value, Ordering::SeqCst, Ordering::Relaxed)
+            .is_err()
+        {
+            let c = &self.cells[base as usize % self.cells.len()];
```

Comment:
Actually, thinking some more on this, I wonder if this will just make all the threads that fail to increment `base` then decide to use the _same_ cell. We could use `base + value`, but even that's not great since `value` is _usually_ -1 or 1. I wonder if there's something else we can seed it based off of. I also wonder if we should change the index on each iteration (currently it ends up being loop-invariant). Maybe by mixing in the value we read from `cv`?

Reply:
I think that'd be a great thing to try and port! But for the purposes of this PR, that _may_ be overkill unless you specifically want to do it :) In the absence of that though, we do need to use a better index than _just_ `base` since all the contending threads will pick the same cell. I think `base + value` that also gets `+= cv` on each loop is a good place to start.